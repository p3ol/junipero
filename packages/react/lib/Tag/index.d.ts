import {
  ReactNode,
  ComponentPropsWithRef,
  ElementType,
  MutableRefObject,
} from 'react';

export declare type TagRef = {
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
};

export declare interface TagProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  className?: string;
  tag?: string | ElementType;
  onDelete?: () => void;
  ref?: MutableRefObject<TagRef | undefined>;
}

declare function Tag(props: TagProps): ReactNode | JSX.Element;

export default Tag;
