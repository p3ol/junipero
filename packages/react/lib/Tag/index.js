import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';
import { forwardRef, useImperativeHandle, useRef } from 'react';

import { Remove } from '../icons';

const Tag = forwardRef(({
  className,
  children,
  onDelete,
  tag: Comp = 'span',
  ...rest
}, ref) => {
  const innerRef = useRef();

  useImperativeHandle(ref, () => ({
    isJunipero: true,
    innerRef,
  }));

  const onDelete_ = e => {
    e.stopPropagation();
    onDelete?.();
  };

  const content = onDelete ? <span>{ children }</span> : children;

  return (
    <Comp
      { ...rest }
      ref={innerRef}
      className={classNames('junipero', 'tag', className)}
    >
      { content }
      { onDelete && (
        <Remove onClick={onDelete_} />
      ) }
    </Comp>
  );
});

Tag.displayName = 'Tag';
Tag.propTypes = {
  tag: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.object,
    PropTypes.func,
  ]),
  onDelete: PropTypes.func,
};

export default Tag;
