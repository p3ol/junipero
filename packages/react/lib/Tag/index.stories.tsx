import { action } from '@storybook/addon-actions';
import { useState } from 'react';

import Tag from './index';

export default { title: 'react/Tag' };

export const basic = () => (
  <>
    <Tag>Default</Tag>
    <Tag className="primary">Primary</Tag>
    <Tag className="danger">Danger</Tag>
    <Tag className="warning">Warning</Tag>
    <Tag className="success">Success</Tag>
  </>
);

export const smaller = () => (
  <>
    <Tag className="info">Default</Tag>
    <Tag className="info primary">Primary</Tag>
    <Tag className="info danger">Danger</Tag>
    <Tag className="info warning">Warning</Tag>
    <Tag className="info success">Success</Tag>
  </>
);

export const deletable = () => {
  const [items, setItems] = useState(['Item 1', 'Item 2', 'Item 3']);

  const onDelete = (item: string) => {
    action('onDelete')(item);
    setItems(items.filter(i => i !== item));
  };

  return items.map(item => (
    <Tag key={item} onDelete={onDelete.bind(null, item)}>
      { item }
    </Tag>
  ));
};
