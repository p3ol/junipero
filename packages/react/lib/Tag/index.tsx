import { classNames } from '@junipero/core';
import PropTypes from 'prop-types';
import { ComponentPropsWithRef, ElementType, MutableRefObject, ReactNode, forwardRef, useImperativeHandle, useRef } from 'react';

import { Remove } from '../icons';
import { ForwardedProps } from '../utils';

export declare type TagRef = {
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
};

export declare interface TagProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  className?: string;
  tag?: string | ElementType;
  onDelete?: () => void;
  ref?: MutableRefObject<TagRef | undefined>;
}
const Tag = forwardRef(({
  className,
  children,
  onDelete,
  tag: Comp = 'span',
  ...rest
}: TagProps, ref) => {
  const innerRef = useRef();

  useImperativeHandle(ref, () => ({
    isJunipero: true,
    innerRef,
  }));

  const onDelete_ = e => {
    e.stopPropagation();
    onDelete?.();
  };

  const content = onDelete ? <span>{ children }</span> : children;

  return (
    <Comp
      { ...rest }
      ref={innerRef}
      className={classNames('junipero', 'tag', className)}
    >
      { content }
      { onDelete && (
        <Remove onClick={onDelete_} />
      ) }
    </Comp>
  );
}) as ForwardedProps<TagProps, TagRef>;

Tag.displayName = 'Tag';
Tag.propTypes = {
  tag: PropTypes.oneOfType([
    PropTypes.any, //TODO fix me
  ]),
  onDelete: PropTypes.func,
};

export default Tag;
