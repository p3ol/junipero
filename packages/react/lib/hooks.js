import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
