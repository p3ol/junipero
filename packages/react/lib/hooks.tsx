import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
