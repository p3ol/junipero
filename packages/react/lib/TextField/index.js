import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
