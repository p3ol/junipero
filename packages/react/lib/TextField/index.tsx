import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
