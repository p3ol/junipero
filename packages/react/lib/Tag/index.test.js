import { createRef, useState } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Tag from './index';

describe('<Tag />', () => {
  it('should render', () => {
    const ref = createRef();
    const { unmount, container } = render(<Tag ref={ref}>New</Tag>);
    expect(container).toMatchSnapshot();
    expect(ref.current.isJunipero).toBe(true);
    expect(ref.current.innerRef.current).toBe(container.firstChild);
    unmount();
  });

  it('should be deletable', async () => {
    const Comp = () => {
      const [items, setItems] = useState(['Item 1', 'Item 2', 'Item 3']);

      const onDelete = item => {
        setItems(items.filter(i => i !== item));
      };

      return items.map(item => (
        <Tag key={item} onDelete={onDelete.bind(null, item)}>{ item }</Tag>
      ));
    };

    const user = userEvent.setup();
    const { unmount, container } = render(<Comp />);
    expect(container).toMatchSnapshot();

    await user.click(container.querySelector('.tag:first-child .delete'));
    expect(container).toMatchSnapshot();

    unmount();
  });
});
