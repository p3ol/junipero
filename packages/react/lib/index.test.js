import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
