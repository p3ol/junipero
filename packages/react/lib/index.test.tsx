import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
