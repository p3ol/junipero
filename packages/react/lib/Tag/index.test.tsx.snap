// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Tag /> should be deletable 1`] = `
<div>
  <span
    class="junipero tag"
  >
    <span>
      Item 1
    </span>
    <svg
      class="junipero icon remove"
      height="10"
      viewBox="0 0 9 10"
      width="10"
    >
      <path
        d="M8 1.5L1 8.5"
      />
      <path
        d="M1 1.5L8 8.5"
      />
    </svg>
  </span>
  <span
    class="junipero tag"
  >
    <span>
      Item 2
    </span>
    <svg
      class="junipero icon remove"
      height="10"
      viewBox="0 0 9 10"
      width="10"
    >
      <path
        d="M8 1.5L1 8.5"
      />
      <path
        d="M1 1.5L8 8.5"
      />
    </svg>
  </span>
  <span
    class="junipero tag"
  >
    <span>
      Item 3
    </span>
    <svg
      class="junipero icon remove"
      height="10"
      viewBox="0 0 9 10"
      width="10"
    >
      <path
        d="M8 1.5L1 8.5"
      />
      <path
        d="M1 1.5L8 8.5"
      />
    </svg>
  </span>
</div>
`;

exports[`<Tag /> should be deletable 2`] = `
<div>
  <span
    class="junipero tag"
  >
    <span>
      Item 1
    </span>
    <svg
      class="junipero icon remove"
      height="10"
      viewBox="0 0 9 10"
      width="10"
    >
      <path
        d="M8 1.5L1 8.5"
      />
      <path
        d="M1 1.5L8 8.5"
      />
    </svg>
  </span>
  <span
    class="junipero tag"
  >
    <span>
      Item 2
    </span>
    <svg
      class="junipero icon remove"
      height="10"
      viewBox="0 0 9 10"
      width="10"
    >
      <path
        d="M8 1.5L1 8.5"
      />
      <path
        d="M1 1.5L8 8.5"
      />
    </svg>
  </span>
  <span
    class="junipero tag"
  >
    <span>
      Item 3
    </span>
    <svg
      class="junipero icon remove"
      height="10"
      viewBox="0 0 9 10"
      width="10"
    >
      <path
        d="M8 1.5L1 8.5"
      />
      <path
        d="M1 1.5L8 8.5"
      />
    </svg>
  </span>
</div>
`;

exports[`<Tag /> should render 1`] = `
<div>
  <span
    class="junipero tag"
  >
    New
  </span>
</div>
`;
